parsers/BG.py

```diff
@@ -1,72 +1,69 @@
+from electricitymap.contrib.lib.types import ZoneKey
+from electricitymap.contrib.parsers.lib.exceptions import ParserException
+
+PARSER = "BG.py"
+
+SOURCE = "eso.bg"
+SOURCE_API_URL = "http://www.eso.bg/api/rabota_na_EEC_json.php?en"
+
+PRODUCTION_TYPE_TO_PRODUCTION_MODE = {
+    "NPP": "nuclear",
+    "CHP": "coal",
+    "Heating TPPs": "gas",
+    "Factory TPPs": "gas",
+    "HPP": "hydro",
+    "Small HPPs": "hydro",
+    "Wind": "wind",
+    "PV": "solar",
+    "Bio": "biomass",
 }
 
 
 def fetch_production(
-    zone_key: str = "BG",
+    zone_key: ZoneKey = ZoneKey("BG"),
     session: Session | None = None,
-    target_datetime=None,
+    target_datetime: datetime | None = None,
     logger: Logger = getLogger(__name__),
-) -> dict:
-    """Requests the last known production mix (in MW) of a given country."""
-    if target_datetime:
-        raise NotImplementedError("This parser is not yet able to parse past dates")
-
-    r = session or Session()
-    url = "http://www.eso.bg/api/rabota_na_EEC_json.php"
-    res = r.get(url)
-
-    assert (
-        res.status_code == 200
-    ), f"Exception when fetching production for {zone_key}: error when calling url={url}"
-
-    response = res.json()
-
-    logger.debug(f"Raw generation breakdown: {response}")
-
-    datapoints = []
-    for row in response:
-        for k in TYPE_MAPPING:
-            if row[0].startswith(k):
-                datapoints.append((TYPE_MAPPING[k], row[1]))
-                break
-
-    production = {}
-    for k, v in datapoints:
-        production[k] = production.get(k, 0.0) + v
-
-    data = {
-        "zoneKey": zone_key,
-        "production": production,
-        "storage": {},
-        "source": "eso.bg",
-        "datetime": datetime.now(timezone.utc),
-    }
-
-    return data
-
-
-if __name__ == "__main__":
-    """Main method, never used by the Electricity Map backend, but handy for testing."""
-
-    print("fetch_production() ->")
-    print(fetch_production())
+) -> list[dict]:
+    session = session or Session()
+
+    if target_datetime is not None:
+        raise ParserException(
+            PARSER, "This parser is not yet able to parse historical data", zone_key
+        )
+
+    time = datetime.now(timezone.utc)
+    response = session.get(SOURCE_API_URL)
+    if not response.ok:
+        raise ParserException(
+            PARSER,
+            f"Exception when fetching production error code: {response.status_code}: {response.text}",
+            zone_key,
+        )
+
+    response_payload = response.json()
+    logger.debug(f"Raw generation breakdown: {response_payload}")
+
+    production_mix = ProductionMix()
+    for header, value in response_payload:
+        production_type, _delimiter, _percentage = header.rpartition(" ")
+        production_mode = PRODUCTION_TYPE_TO_PRODUCTION_MODE.get(production_type)
+        if production_mode in PRODUCTION_MODES:
+            production_mix.add_value(production_mode, value)
```

Comment:
We should skip this check, if there is a mode that is not a part of the production modes (which is checked internally) we want to raise logs for it so we can map it to something valid.